File: ayeon/행렬테두리회전하기.cpp
// P
#include <string>
#include <vector>

using namespace std;

vector<int> solution(int rows, int columns, vector<vector<int>> queries) {
    vector<int> answer;
    
    vector<vector<int>> vec(rows, vector<int>(columns));
    
    int k=1;
    
    for (int i = 0; i < vec.size(); i++)
        for (int j = 0; j < vec[i].size(); j++)
            vec[i][j] = k++; 
    
    for(int i=0; i<queries.size(); i++)
    {
        int x1,y1,x2,y2;
        x1= queries[i][0] -1;
        y1= queries[i][1] -1;
        x2= queries[i][2] -1;
        y2= queries[i][3] -1;
        
        int s= vec[x1][y1];
        
        int min=s;
        int next;
        
        for(int i=x1; i<x2; i++){
            next = vec[i+1][y1];
    	    vec[i][y1] = next;
            if(min>next) min = next;
        }
        for(int i=y1; i<y2; i++){
            next = vec[x2][i+1];
            vec[x2][i]=next;
            if(min>next) min = next;
        }
        for(int i=x2; i>x1; i--){
            next = vec[i-1][y2];
            vec[i][y2] = next;
            if(min>next) min = next;
        }
        for(int i=y2; i>y1; i--){
            next = vec[x1][i-1];
            vec[x1][i] = next;
            if(min>next) min = next;
        }
        vec[x1][y1+1] = s;
        
        answer.push_back(min);
    }
    
    return answer;
}

File: minnkyoung/행렬 테두리 회전하기.py
def solution(rows, columns, queries):
    answer = []
    board = [[] for _ in range(rows)]
    
    for i in range(1, rows+1):
        for j in range(1, columns+1):
            board[i-1].append((i-1)*columns+j)
    
    for x1, y1, x2, y2 in queries:
        temp = board[x1-1][y2-1]
        min_value = 10001
        
        # 북쪽 테두리
        min_value = min(min(board[x1-1][y1-1:y2-1]), min_value)
        board[x1-1][y1:y2] = board[x1-1][y1-1:y2-1]
        
        # 서쪽 테두리
        for i in range(x1, x2):
            min_value = min(board[i][y1-1], min_value)
            board[i-1][y1-1] = board[i][y1-1]
        
        # 남쪽 테두리
        min_value = min(min(board[x2-1][y1:y2]), min_value)
        board[x2-1][y1-1:y2-1] = board[x2-1][y1:y2]
        
        # 동쪽 테두리
        for i in range(x2-2, x1-2, -1):
            min_value = min(board[i][y2-1], min_value)
            board[i+1][y2-1] = board[i][y2-1]
        
        board[x1][y2-1] = temp
        min_value = min(min_value, temp)
        
        answer.append(min_value)
    
    return answer